### ISSUE TITLE
ext_pillar gitfs: Unable to checkout branch None

### ISSUE THREAD

**Issue by dhs-rec** [saltstack/salt#23842]
I've got a problem with git based external pillar in 2015.5.0. I've setup two external pillars, one based on MongoDB, the other one based on git. When looking at pillar items for a given minion ("salt minion pillar.items"), I only get the ones from the MongoDB pillar (and from the masters filesystem), but not the ones from git.

I have the following lines in the master config file:

```
ext_pillar:
  - mongo: {}
  - git: __env__ github:organization/salt-test.git root=pillar
```

and both the filesystem and git repository contain a top.sls with this content:

```
{{ env }}:
  '*':
    - data
```

The master log file contains this line:

```
[ERROR   ][31414] Unable to checkout branch None: 'git checkout origin/None' returned exit status 1: error: pathspec 'origin/None' did not match any file(s) known to git.
```

**Comment by rallytime** [comment]
@dhs-rec Can you post the output of `salt --versions-report`, please? Also, I am not terribly familiar with the inner workings of ext_pillar, so I am not quite sure what you're doing with the `__env__` var you're passing in there, but have you looked at the docs for [git_pillar](http://docs.saltstack.com/en/latest/ref/pillar/all/salt.pillar.git_pillar.html)? That might give you some pointers.

**Comment by dhs-rec** [comment]
Versions:
```
# salt --versions-report
                  Salt: 2015.5.0
                Python: 2.7.6 (default, Mar 22 2014, 22:59:56)
                Jinja2: 2.7.2
              M2Crypto: 0.21.1
        msgpack-python: 0.3.0
          msgpack-pure: Not Installed
              pycrypto: 2.6.1
               libnacl: Not Installed
                PyYAML: 3.10
                 ioflo: Not Installed
                 PyZMQ: 14.0.1
                  RAET: Not Installed
                   ZMQ: 4.0.4
                  Mako: 0.9.1
 Debian source package: 2015.5.0+ds-1trusty1
```
Yes, I've looked at those docs, that's the reason why I did this setup in the first place. Oh, talking about the docs: I think there's an error in the new layout. Near the top of the page, it says:
> Changed in version 2015.5.0: The special branch name '`__env__`' will be replace by the environment ({{env}})

Then later, the example looks like:
```
ext_pillar:
  - git: _ git://gitserver/git-pillar.git root=pillar
```
Looks like the `__env__` (as it looked like in the old layout) has turned into an single `_` in the new layout...

**Comment by rallytime** [comment]
Oh ok. So do you just consider this a doc fix? Or just something that changed when you upgraded, but is working now?

**Comment by dhs-rec** [comment]
No, the doc thing is just a side note.

**Comment by rallytime** [comment]
@whiteinge Do you know anything about this change or maybe @UtahDave?

**Comment by whiteinge** [comment]
The correct argument is `__env__` and the docs have been fixed in develop but not 2015.2 (need to backport ##22572). The error message makes it look like that variable isn't getting swapped out correctly  so this sounds like a regression.

**Comment by rallytime** [comment]
I've backported the doc fix in #24057 and added the regression label.

**Comment by dhs-rec** [comment]
Is there any news on this issue?

**Comment by keesbos** [comment]
The '_' was replaced by '__env__'.
I found out that somewhere in 2015.5.x the update frequency changed. Which was generally good, but not for the '__env__' handling.

So,
```yaml
ext_pillar:
  - git: __env__ git://gitserver/git-pillar.git root=pillar
```
should be the correct usage for the legacy config.

Side note: This being legacy git_pillar is new for me. I noticed that the new git_pillar code uses pillarenv. This seems to be the same as saltenv for states gitfs (i.e. setting the root fixed). If this is true, that will make __env__ obsolete (i.e. is more or less doing the same). However, I cannot see it working in the latest release (2015.08.1) as I would expect from looking at the code (pillarenv is allways None). I'll drop a note here, when I know whether the pillarenv will replace __env__ or not.

**Comment by peter-slovak** [comment]
If you use the `__env__` keyword in Git pillar config, make sure you statically set the `environment` minion option to the desired environment/branch on your minions. Without it, `__env__` resolves to None. This worked fine for me until 2015.8.0 where the external pillar was rewritten and using `__env__` now tries to checkout a branch with that literal name. See #28125

**Comment by terminalmage** [comment]
https://github.com/saltstack/salt/pull/36659 implements ``__env__`` support in the re-written git_pillar. As for legacy git_pillar, I am unable to reproduce this. It seems that this was likely fixed in 6dedeac, which was merged about a week and a half before this issue was originally opened, so the fix likely just hadn't propagated to an actual release at that time.
